[css-lists] Implement `counter-set` property

nt1m opened the pull request:
#### 35c672ab42641db0295b649a694a3355bf2b03e2
<pre>
[css-lists] Implement `counter-set` property
<a href="https://bugs.webkit.org/show_bug.cgi?id=97337">https://bugs.webkit.org/show_bug.cgi?id=97337</a>
rdar://88663348

Reviewed by Cameron McCormack and Darin Adler.

This implements the `counter-set` property as per <a href="https://drafts.csswg.org/css-lists/#propdef-counter-set">https://drafts.csswg.org/css-lists/#propdef-counter-set</a> .

This property applies at the final step after:
- counter-reset
- counter-increment

Unlike counter-reset, counter-set does not create new counters if they already exist.

In order for interactions with the `list-item` counter to be correct, it does these changes per the HTML spec:
- It starts using a `counter-set` presentional hint to implement the `&lt;li&gt;` value attribute
- `counter-reset`, `counter-set` and `counter-increment` can now be overriden on list-item

This matches Firefox&apos;s behavior and is also closer to Chrome&apos;s.

* LayoutTests/fast/css/counters/counter-list-item-expected.html:
* LayoutTests/fast/css/counters/counter-list-item.html:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/css-lists-no-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/counter-set-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-rounds-to-integer-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:

Rebaselines.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):

Add animation support for counter-set.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::counterToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterSet):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:

Parse and define the counter-set property.

* Source/WebCore/html/HTMLLIElement.cpp:
(WebCore::HTMLLIElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLLIElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLLIElement::didAttachRenderers):
(WebCore::HTMLLIElement::attributeChanged): Deleted.
(WebCore::HTMLLIElement::parseValue): Deleted.
* Source/WebCore/html/HTMLLIElement.h:

Implement &lt;li value&gt; as counter-set presentiational hint.

* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::CounterNode::CounterNode):
(WebCore::CounterNode::create):
(WebCore::CounterNode::computeCountInParent const):
(WebCore::CounterNode::insertAfter):
* Source/WebCore/rendering/CounterNode.h:
(WebCore::CounterNode::actsAsReset const):
(WebCore::CounterNode::hasResetType const):
(WebCore::CounterNode::hasSetType const):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::listItemCounterDirectives):
(WebCore::planCounter):
(WebCore::findPlaceForCounter):
(WebCore::makeCounterNode):

Integrate counter-set with counters.

* Source/WebCore/rendering/RenderListItem.cpp:
(WebCore::RenderListItem::updateValueNow const):
(WebCore::RenderListItem::updateValue):

Take in account counter-set / counter-increment when calculating the value.

(WebCore::RenderListItem::styleDidChange):
(WebCore::RenderListItem::counterDirectivesChanged):

Invalidate list markers when counter properties change.

(WebCore::RenderListItem::explicitValueChanged): Deleted.
(WebCore::RenderListItem::setExplicitValue): Deleted.

Delete old &lt;li value&gt; code, replaced by presentational hints.

* Source/WebCore/rendering/RenderListItem.h:
* Source/WebCore/rendering/style/CounterDirectives.h:
(WebCore::operator==):

Expand equality operator.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritCounter):
(WebCore::Style::BuilderCustom::applyValueCounter):
(WebCore::Style::BuilderCustom::applyInitialCounterSet):
(WebCore::Style::BuilderCustom::applyInheritCounterSet):
(WebCore::Style::BuilderCustom::applyValueCounterSet):

Convert CSSValues to counter directives.

Canonical link: <a href="https://commits.webkit.org/267137@main">https://commits.webkit.org/267137@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/446a434c2cd53366013a932dc983506844b08a9f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16292 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17373 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14652 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15799 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17188 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15802 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16267 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13292 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18118 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14094 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21021 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14546 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14261 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17525 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14848 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12589 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14104 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13941 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3767 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18466 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14668 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderListItem.cpp:
```diff
@@ -231,25 +231,47 @@ void RenderListItem::updateValueNow() const
     }
 
+    int defaultIncrement = (orderedList && orderedList->isReversed()) ? -1 : 1;
+    // FIXME: This should account for counter-reset as well.
```
What tests show this omission causes a bug? Do we have some?

nt1m replied:
There are some tests on WPT, but also our own layout test seems to exercise this (although the expected results are incorrect). I'll make a basic fix so that test passes, but I'll do the full support in a follow up.

darinadler commented on Source/WebCore/rendering/RenderListItem.cpp:
```diff
@@ -231,25 +231,47 @@ void RenderListItem::updateValueNow() const
+{
+    RenderBlockFlow::styleDidChange(diff, oldStyle);
+
+    if (!oldStyle || (oldStyle->counterDirectives().map.isEmpty() && style().counterDirectives().map.isEmpty()))
```
Why do we need this special case for when both maps are empty? Seems like the code below already will return efficiently in that case.

darinadler commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -170,7 +171,7 @@ class BuilderCustom {
     static void applyTextOrBoxShadowValue(BuilderState&, CSSValue&);
 
-    enum CounterBehavior {Increment = 0, Reset};
+    enum CounterBehavior { Increment = 0, Reset, Set };
```
We don’t need that `= 0` and I would remove it.

darinadler reviewed (approved):
Let's make sure this doesn't measurably slow down handling of `li` elements with values.

heycam commented on Source/WebCore/rendering/RenderListItem.cpp:
```diff
@@ -231,25 +231,47 @@ void RenderListItem::updateValueNow() const
         }
     }
 
+    int defaultIncrement = (orderedList && orderedList->isReversed()) ? -1 : 1;
```
Nit: parens around the LHS of the ternary operator not needed.

heycam commented on Source/WebCore/rendering/RenderListItem.cpp:
```diff
@@ -231,25 +231,47 @@ void RenderListItem::updateValueNow() const
         item = nextListItem(*list, *item);
-        item->m_value = (value += increment);
+        auto directives = item->style().counterDirectives().map.get("list-item"_s);
+        item->m_value = (value = valueForItem(value, directives));
```
Please separate these out into two statements.

heycam commented on Source/WebCore/rendering/RenderListItem.cpp:
```diff
@@ -231,25 +231,47 @@ void RenderListItem::updateValueNow() const
-    if (!startValue)
-        startValue = orderedList ? orderedList->start() : 1;
+    if (!startValue) {
+        auto directives = startItem->style().counterDirectives().map.get("list-item"_s);
```
This copies the `CounterDirectives` object. Though at 16 bytes it's not a big deal.

nt1m replied:
will use `auto&` if possible.

heycam replied:
I was trying to read the `HashMap` and `HashTraits` code to work out if `get()` returned a referenced or a copy of the data, but wasn't sure without building.

darinadler replied:
The `get` function returns a copy, not a reference to the value in the map, so it can handle values not in the map. If you want to avoid a copy you have to use the wordier `find`/`end` combination.

darinadler replied:
I think that in this case a copy may be a good thing. A small scalar that likely ends up in registers.

heycam commented on Source/WebCore/rendering/RenderListItem.cpp:
```diff
@@ -231,25 +231,47 @@ void RenderListItem::updateValueNow() const
+{
+    RenderBlockFlow::styleDidChange(diff, oldStyle);
+
+    if (!oldStyle || (oldStyle->counterDirectives().map.isEmpty() && style().counterDirectives().map.isEmpty()))
```
Probably not worth checking that these are empty and returning early. Most `RenderListItem`s will have a list-item counter directive.

nt1m replied:
> Most RenderListItems will have a list-item counter directive.

The map wouldn't necessarily be populated given the special implementation of the list-item directive, only the `value` attribute on `<li>` or explicit counter properties would populate it.

heycam commented on Source/WebCore/rendering/RenderCounter.cpp:
```diff
@@ -165,23 +165,33 @@ static CounterDirectives listItemCounterDirectives(RenderElement& renderer)
-            return { list.start(), list.isReversed() ? 1 : -1 };
+            return {
+                .resetValue = list.start(),
+                .incrementValue = list.isReversed() ? 1 : -1,
```
I'm confused by this returning of 1 if the list is reversed (which is in the existing code too). Why do we do this?

nt1m replied:
Probably a typo, I'll try to add a testcase for this.

nt1m replied:
To my surprise, this is actually correct, there's some strange ordering in LayoutTests/fast/css/counters/counter-list-item.html otherwise.

heycam commented on Source/WebCore/html/HTMLLIElement.cpp:
```diff
@@ -89,19 +89,13 @@ void HTMLLIElement::collectPresentationalHintsForAttribute(const QualifiedName&
         }
+    } else if (name == valueAttr) {
+        if (auto parsedValue = parseHTMLInteger(value))
+            addPropertyToPresentationalHintStyle(style, CSSPropertyCounterSet, makeString("list-item "_s, *parsedValue));
```
There's an extra space in the string here. That would cause `value=""` not to work at all, right?

If this is not causing test failures, please add a test that would've failed.

nt1m replied:
You want it to be `counter-set: list-item 4` not `counter-set: list-item4`. `imported/w3c/web-platform-tests/css/css-lists/li-value-counter-reset-001.html` would have failed otherwise.

`value=""` would just fallback to normal rendering as `parseHTMLInteger` would give a parse error. See: https://html.spec.whatwg.org/#the-li-element

heycam replied:
Of course, you are right.

nt1m committed:
[css-lists] Implement `counter-set` property
https://bugs.webkit.org/show_bug.cgi?id=97337
rdar://88663348

Reviewed by Cameron McCormack and Darin Adler.

This implements the `counter-set` property as per https://drafts.csswg.org/css-lists/#propdef-counter-set .

This property applies at the final step after:
- counter-reset
- counter-increment

Unlike counter-reset, counter-set does not create new counters if they already exist.

In order for interactions with the `list-item` counter to be correct, it does these changes per the HTML spec:
- It starts using a `counter-set` presentional hint to implement the `<li>` value attribute
- `counter-reset`, `counter-set` and `counter-increment` can now be overriden on list-item

This matches Firefox's behavior and is also closer to Chrome's.

* LayoutTests/fast/css/counters/counter-list-item-expected.html:
* LayoutTests/fast/css/counters/counter-list-item.html:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/css-lists-no-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/counter-set-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-rounds-to-integer-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:

Rebaselines.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):

Add animation support for counter-set.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::counterToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterSet):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:

Parse and define the counter-set property.

* Source/WebCore/html/HTMLLIElement.cpp:
(WebCore::HTMLLIElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLLIElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLLIElement::didAttachRenderers):
(WebCore::HTMLLIElement::attributeChanged): Deleted.
(WebCore::HTMLLIElement::parseValue): Deleted.
* Source/WebCore/html/HTMLLIElement.h:

Implement <li value> as counter-set presentiational hint.

* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::CounterNode::CounterNode):
(WebCore::CounterNode::create):
(WebCore::CounterNode::computeCountInParent const):
(WebCore::CounterNode::insertAfter):
* Source/WebCore/rendering/CounterNode.h:
(WebCore::CounterNode::actsAsReset const):
(WebCore::CounterNode::hasResetType const):
(WebCore::CounterNode::hasSetType const):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::listItemCounterDirectives):
(WebCore::planCounter):
(WebCore::findPlaceForCounter):
(WebCore::makeCounterNode):

Integrate counter-set with counters.

* Source/WebCore/rendering/RenderListItem.cpp:
(WebCore::RenderListItem::updateValueNow const):
(WebCore::RenderListItem::updateValue):

Take in account counter-set / counter-increment when calculating the value.

(WebCore::RenderListItem::styleDidChange):
(WebCore::RenderListItem::counterDirectivesChanged):

Invalidate list markers when counter properties change.

(WebCore::RenderListItem::explicitValueChanged): Deleted.
(WebCore::RenderListItem::setExplicitValue): Deleted.

Delete old <li value> code, replaced by presentational hints.

* Source/WebCore/rendering/RenderListItem.h:
* Source/WebCore/rendering/style/CounterDirectives.h:
(WebCore::operator==):

Expand equality operator.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritCounter):
(WebCore::Style::BuilderCustom::applyValueCounter):
(WebCore::Style::BuilderCustom::applyInitialCounterSet):
(WebCore::Style::BuilderCustom::applyInheritCounterSet):
(WebCore::Style::BuilderCustom::applyValueCounterSet):

Convert CSSValues to counter directives.

Canonical link: https://commits.webkit.org/267137@main
